# Practical Application III: Comparing Classifiers

**Overview**: In this practical application, your goal is to compare the performance of the classifiers we encountered in this section, namely K Nearest Neighbor, Logistic Regression, Decision Trees, and Support Vector Machines.  We will utilize a dataset related to marketing bank products over the telephone.  



### Getting Started

Our dataset comes from the UCI Machine Learning repository [link](https://archive.ics.uci.edu/ml/datasets/bank+marketing).  The data is from a Portugese banking institution and is a collection of the results of multiple marketing campaigns.  We will make use of the article accompanying the dataset [here](CRISP-DM-BANK.pdf) for more information on the data and features.



### Problem 1: Understanding the Data

To gain a better understanding of the data, please read the information provided in the UCI link above, and examine the **Materials and Methods** section of the paper.  How many marketing campaigns does this data represent?

In [ ]:
# Dataset represents the result of multiple marketing campaigns conducted by Portugeuse banking institution. 
# The campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the product 
# (bank term deposit) would be ('yes') or not ('no') subscribed. 
#The dataset's structure allows us to analyze the effectiveness of these telemarketing efforts on a per-client basis, 
# with each row representing an outcome of a contact with a specific client during the campaigns.

### Problem 2: Read in the Data

Use pandas to read in the dataset `bank-additional-full.csv` and assign to a meaningful variable name.

In [1]:
import pandas as pd

In [6]:
df = pd.read_csv('data/bank-additional-full.csv', sep = ';')

In [9]:
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

### Problem 3: Understanding the Features


Examine the data description below, and determine if any of the features are missing values or need to be coerced to a different data type.


```
Input variables:
# bank client data:
1 - age (numeric)
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
5 - default: has credit in default? (categorical: 'no','yes','unknown')
6 - housing: has housing loan? (categorical: 'no','yes','unknown')
7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:
8 - contact: contact communication type (categorical: 'cellular','telephone')
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
# other attributes:
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
14 - previous: number of contacts performed before this campaign and for this client (numeric)
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
# social and economic context attributes
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
17 - cons.price.idx: consumer price index - monthly indicator (numeric)
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')
```



In [7]:

# populate the bank_data variable with the contents of the CSV file
bank_data = df

# Display the first few rows of the DataFrame to confirm it was read correctly
print(bank_data.head())

# Check for missing values
missing_values = bank_data.isnull().sum()
print("\nMissing values in each column:\n", missing_values)

# Check the data types of each column
data_types = bank_data.dtypes
print("\nData types of each column:\n", data_types)

   age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed

### Problem 4: Understanding the Task

After examining the description and data, your goal now is to clearly state the *Business Objective* of the task.  State the objective below.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

**Business Objective**

Key business objective of this task is to develop a predictive model that can accurately forecast whether a client will subscribe to term depositi based on historical marketing data. By leveraging this mode, the bank would aim to optimize its telmarketing strategies, improve targeting efficiency, and increae subscription date of the term deposit.

The success of this project will be measured by the model's ability to accurately predict the target variable (y), which indicates whether or not a client subscribed to a term deposit, while providing actionable insights to improve future marketing campaigns.

### Problem 5: Engineering Features

Now that you understand your business objective, we will build a basic model to get started.  Before we can do this, we must work to encode the data.  Using just the bank information features, prepare the features and target column for modeling with appropriate encoding and transformations.

To prepare data for modeling, we will focus on encoding and transforming the bank information numerical and categoricaly features.

In [10]:
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Initialize the label encoder
label_encoders = {}

# Apply Label Encoding to categorical features
for col in ['job', 'marital', 'education', 'default', 'housing', 
            'loan', 'contact', 'month', 'day_of_week', 'poutcome']:
    le = LabelEncoder()
    bank_data[col] = le.fit_transform(bank_data[col])
    label_encoders[col] = le

# Features to be used in the model (bank information only)
bank_info_features = ['age', 'job', 'marital', 'education', 'default', 'housing', 
                      'loan', 'contact', 'month', 'day_of_week', 'duration', 
                      'campaign', 'pdays', 'previous', 'poutcome']

# Define the preprocessing steps for numerical features only
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), ['age', 'duration', 'campaign', 'pdays', 'previous'])
    ],
    remainder='passthrough'  # This keeps the non-numerical columns as they are
)

# Prepare the features (X) and target (y)
X_bank_info = bank_data[bank_info_features]
y = bank_data['y'].apply(lambda x: 1 if x == 'yes' else 0)

# Apply the transformations to the features
X_bank_info_transformed = preprocessor.fit_transform(X_bank_info)

# Display the shape of the transformed features and target
X_bank_info_transformed.shape, y.shape


((41188, 15), (41188,))

### Problem 6: Train/Test Split

With your data prepared, split it into a train and test set.

In [12]:
from sklearn.model_selection import train_test_split

# Perform the train/test split (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X_bank_info_transformed, y, test_size=0.2, random_state=42, stratify=y)

# Display the shapes of the train and test sets
print("Training set shape:", X_train.shape, y_train.shape)
print("Testing set shape:", X_test.shape, y_test.shape)


Training set shape: (32950, 15) (32950,)
Testing set shape: (8238, 15) (8238,)


### Problem 7: A Baseline Model

Before we build our first model, we want to establish a baseline.  What is the baseline performance that our classifier should aim to beat?

In [13]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score

# Create a DummyClassifier that uses the 'most_frequent' strategy
dummy_clf = DummyClassifier(strategy='most_frequent')

# Fit the dummy classifier on the training data
dummy_clf.fit(X_train, y_train)

# Predict using the dummy classifier on the test data
y_pred_dummy = dummy_clf.predict(X_test)

# Calculate the accuracy of the dummy classifier
baseline_accuracy = accuracy_score(y_test, y_pred_dummy)

print("Baseline accuracy (most frequent class):", baseline_accuracy)

Baseline accuracy (most frequent class): 0.8873512988589464


### Problem 8: A Simple Model

Use Logistic Regression to build a basic model on your data.  

In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Create a Logistic Regression model
log_reg = LogisticRegression(max_iter=1000, random_state=42)

# Train the model on the training data
log_reg.fit(X_train, y_train)

# Make predictions on the test data
y_pred = log_reg.predict(X_test)

# Evaluate the model's performance

conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print("\nConfusion Matrix:\n", conf_matrix)
print("\nClassification Report:\n", class_report)


Confusion Matrix:
 [[7171  139]
 [ 605  323]]

Classification Report:
               precision    recall  f1-score   support

           0       0.92      0.98      0.95      7310
           1       0.70      0.35      0.46       928

    accuracy                           0.91      8238
   macro avg       0.81      0.66      0.71      8238
weighted avg       0.90      0.91      0.90      8238



### Problem 9: Score the Model

What is the accuracy of your model?

In [16]:
# get accuracy of the model 
accuracy = accuracy_score(y_test, y_pred)

print("Logistic Regression Model Accuracy:", accuracy)

Logistic Regression Model Accuracy: 0.909686817188638


### Problem 10: Model Comparisons

Now, we aim to compare the performance of the Logistic Regression model to our KNN algorithm, Decision Tree, and SVM models.  Using the default settings for each of the models, fit and score each.  Also, be sure to compare the fit time of each of the models.  Present your findings in a `DataFrame` similar to that below:

| Model | Train Time | Train Accuracy | Test Accuracy |
| ----- | ---------- | -------------  | -----------   |
|     |    |.     |.     |

In [22]:
import pandas as pd
import time
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# Initialize models
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "KNN": KNeighborsClassifier(),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "SVM": SVC(random_state=42)
}

# Prepare a DataFrame to store the results
results = pd.DataFrame(columns=["Model", "Train Time", "Train Accuracy", "Test Accuracy"])

# Train and evaluate each model
for model_name, model in models.items():
    start_time = time.time()
    model.fit(X_train, y_train)
    train_time = time.time() - start_time
    
    train_accuracy = accuracy_score(y_train, model.predict(X_train))
    test_accuracy = accuracy_score(y_test, model.predict(X_test))
    print(f"{model_name} -- Train Accuracy: {train_accuracy:.4f}, Test Accuracy: {test_accuracy:.4f}, Train Time: {train_time:.4f} seconds")
    


Logistic Regression -- Train Accuracy: 0.9047, Test Accuracy: 0.9097, Train Time: 0.2186 seconds
KNN -- Train Accuracy: 0.9212, Test Accuracy: 0.9009, Train Time: 0.0186 seconds
Decision Tree -- Train Accuracy: 1.0000, Test Accuracy: 0.8725, Train Time: 0.0784 seconds
SVM -- Train Accuracy: 0.9026, Test Accuracy: 0.9024, Train Time: 4.1321 seconds


| Model | Train Time | Train Accuracy | Test Accuracy |
| ----- | ---------- | -------------  | -----------   |
|   Logistic Regression  |  0.1225  | 0.9097   | 0.9047|
|   KNN  |  0.0186  | 0.9212   | 0.9009|
|   Decision Tree |  0.0784  | 1.0000   | 0.8725|
|   SVM  |  4.1321  | 0.9026   | 0.9024|


### Problem 11: Improving the Model

Now that we have some basic models on the board, we want to try to improve these.  Below, we list a few things to explore in this pursuit.

- More feature engineering and exploration.  For example, should we keep the gender feature?  Why or why not?
- Hyperparameter tuning and grid search.  All of our models have additional hyperparameters to tune and explore.  For example the number of neighbors in KNN or the maximum depth of a Decision Tree.  
- Adjust your performance metric

## Hyperparameters tunning using GridSeachCV for KNN and decision tree models##

In [27]:
from sklearn.model_selection import GridSearchCV

# Define parameter grids for KNN and Decision Tree
param_grid_knn = {
    'n_neighbors': [3, 5, 7, 9],
    'weights': ['uniform', 'distance']
}

param_grid_dt = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10]
}

# Initialize models
knn = KNeighborsClassifier()
dt = DecisionTreeClassifier(random_state=42)

# Initialize GridSearchCV for each model
grid_search_knn = GridSearchCV(knn, param_grid_knn, cv=5, scoring='accuracy')
grid_search_dt = GridSearchCV(dt, param_grid_dt, cv=5, scoring='accuracy')

# Fit GridSearchCV for each model
grid_search_knn.fit(X_train, y_train)
grid_search_dt.fit(X_train, y_train)

# Get the best parameters and best score for each model
best_params_knn = grid_search_knn.best_params_
best_score_knn = grid_search_knn.best_score_

best_params_dt = grid_search_dt.best_params_
best_score_dt = grid_search_dt.best_score_

print("Best KNN Parameters:", best_params_knn)
print("Best KNN Cross-Validation Accuracy:", best_score_knn)

print("Best Decision Tree Parameters:", best_params_dt)
print("Best Decision Tree Cross-Validation Accuracy:", best_score_dt)

Best KNN Parameters: {'n_neighbors': 9, 'weights': 'uniform'}
Best KNN Cross-Validation Accuracy: 0.8996054628224582
Best Decision Tree Parameters: {'max_depth': 10, 'min_samples_split': 10}
Best Decision Tree Cross-Validation Accuracy: 0.8998179059180575


##### Questions